make Shelf a regular object with an `_ingredients` attribute, instead a subclass of `dict`

radix opened the pull request:
There were two problems with `Shelf.__init__`:

- keyword arguments like `metadata`, `table`, `select_from` were buggy -- they would be passed on to the dict constructor before being popped out of `kwargs`, so you would end up with unexpected "ingredients" in the Shelf dictionary named `metadata`, `table`, or `select_from`. This would often then immediately blow up at the end of `Shelf.__init__` when it iterated over those non-ingredients and tried to assign the `.id` attribute on them.

- the `Meta` attribute was just a nested class definition on the `Shelf` class. There is only one of them, but each time `Shelf.__init__` is called, it would overwrite some of the attributes.

  This belies a more complex issue, which is that there is no place to store any extra data on an instance that subclasses `AttrDict`, since it subclasses dict and then sets `self.__dict__ = self`, meaning there is no actual attribute dictionary on this object. Before I saw the `self.__dict__ = self` line, the naive thing I tried was to just say `self.Meta = type(self).Meta()` (to instantiate a new copy of the `Meta` class for each `Shelf` instance), but of course this just stored `Meta` as an ingredient on self, because of `__setattr__` and the fact that there is no separate ingredient storage from Shelf attribute storage.

To solve these problems, I changed `Shelf` so it no longer subclasses `AttrDict`, and instead now store ingredients in a plain dictionary at `Shelf._ingredients`. I then implemented the remainder of the `dict` interface that the recipe test suite was relying on in terms of `self._ingredients`. I removed the `__getattr__` and `__setattr__` implementations because they don't seem to actually be used by apps that use recipe.

I also got rid of the mutation of ingredients `id` in `__getitem__` and `__init__`, so that Shelves no longer mutate things that it may not "own" (e.g. have a copy of). Now all mutation goes through `__setitem__`, which copies and annotates the ingredients, so `__getitem__` doesn't need to mutate the ingredients on the way out. *However*, `__getitem__` still sets `anonymize` because that's something that may change over time on the Shelf. Would be nice to have a better mechanism for this in the future.

chrisgemignani commented on recipe/shelf.py:
```diff
@@ -427,30 +432,55 @@ class Meta:
         select_from = None
         engine = None
         ingredient_order = []
+        metadata = None
```
shelf.Meta.metadata is for the recipe from config?

radix replied:
oh, no. I think I misunderstood some other code as passing metadata to the Shelf constructor. I'll get rid of this :)

radix replied:
wait, no. I added it here because the code was already setting self.Meta.metadata

chrisgemignani commented on recipe/shelf.py:
```diff
@@ -427,30 +432,55 @@ class Meta:
-
+        self.Meta = type(self).Meta()
         self.Meta.ingredient_order = []
         self.Meta.table = kwargs.pop('table', None)
```
add engine in here?

```self.Meta.engine = kwargs.pop('engine', None)```

radix replied:
nothing seems to actually use the Meta.engine attribute, so I would rather just delete that. thoughts?

chrisgemignani replied:
Lets do it. Can't remember why it was there. We'll add it back in if it was needed.

radix commented:
I believe I've addressed your comment (though I think I want to delete the `engine = None` from `class Meta` still, pending your comment). 

I've done one additional thing: changed __getitem__ to not mutate the ingredients on the way out. Previously it would be possible to share ingredients between shelves accidentally, then when accessing them through either shelf, their `id` and `metadata` attributes would be mutated, leading to race conditions.

Instead, the code now reliably always copies & annotates the ingredients on the way *in*, so there should no longer be race conditions.
